sources/core/Stride.Core/Threading/Dispatcher.cs

```diff
@@ -25,497 +25,318 @@ public class Dispatcher
-                }
-                else
-                {
-                    var state = BatchState.Acquire();
-                    state.WorkDone = state.StartInclusive = fromInclusive;
+            // Performs 1/8 to 1/4 better in most cases, performs up to 1/8 worse when the ratio between
+            // the duration each individual item takes and the amount of items per batch hits a very narrow sweet-spot.
+            // Not entirely sure why yet.
+#if FALSE
+            if (items / MaxDegreeOfParallelism > 8)
+                itemsPerBatch /= 4; // Batches of 2 instead of 8 to allow faster threads to steal more of the work
+            else if (items / MaxDegreeOfParallelism > 4)
+                itemsPerBatch /= 2; // Batches of 2 instead of 4 to allow faster threads to steal more of the work
+#endif
 
-                    try
-                    {
-                        var batchCount = Math.Min(MaxDegreeOfParallelism, count);
-                        var batchSize = (count + (batchCount - 1)) / batchCount;
+            var batch = BatchState<TJob>.Borrow(itemsPerBatch: itemsPerBatch, endExclusive: (uint)items, references: batchCount, batchJob);
+            try
+            {
+                ThreadPool.Instance.QueueUnsafeWorkItem(batch, &TypeAdapter<TJob>, batchCount - 1);
 
-                        // Kick off a worker, then perform work synchronously
-                        state.AddReference();
-                        Fork(toExclusive, batchSize, MaxDegreeOfParallelism, action, state);
+                ProcessBatch(batchJob, batch);
 
-                        // Wait for all workers to finish
-                        state.WaitCompletion(toExclusive);
+                // Might as well steal some work instead of just waiting,
+                // also helps prevent potential deadlocks from badly threaded code
+                while (Volatile.Read(ref batch.ItemsDone) < batch.Total && batch.Finished.WaitOne(0) == false)
+                    ThreadPool.Instance.TryCooperate();
 
-                        var ex = Interlocked.Exchange(ref state.ExceptionThrown, null);
-                        if (ex != null)
-                            throw ex;
-                    }
-                    finally
-                    {
-                        state.Release();
-                    }
-                }
+                var ex = Interlocked.Exchange(ref batch.ExceptionThrown, null);
+                if (ex != null)
+                    throw ex;
+            }
+            finally
+            {
+                batch.Release();
             }
         }
 
-        public static void For<TLocal>(int fromInclusive, int toExclusive, [Pooled] Func<TLocal> initializeLocal, [Pooled] Action<int, TLocal> action, [Pooled] Action<TLocal> finalizeLocal = null)
+        private static void TypeAdapter<TJob>(object obj) where TJob : IBatchJob
         {
-            using (Profile(action))
+            var batch = (BatchState<TJob>)obj;
```

Comment:
Yep, it's guaranteed;
```cs
ThreadPool.Instance.QueueUnsafeWorkItem(batch, &TypeAdapter<TJob>, batchCount - 1);
```
This call schedules the function `void TypeAdapter<TJob>(object)` to run on other threads with `batch` passed as the `obj` parameter. Both `TypeAdapter<TJob>` and `BatchState<TJob>`'s generic types match as they both come from the `ForBatched<TJob>` generic.
It would be a fairly serious bug if there wasn't any guarantee.
But as you said, `as` is faster than direct cast, it won't move the needle that much compared to how long those jobs tend to take, but anything helps. I'll change it to `as` shortly.